services/src/main/java/org/keycloak/services/managers/AuthenticationManager.java

```diff
@@ -215,22 +216,41 @@ public static void backchannelLogout(KeycloakSession session, RealmModel realm,
         }
     }
 
-    private static AuthenticationSessionModel createOrJoinLogoutSession(RealmModel realm, final AuthenticationSessionManager asm, boolean browserCookie) {
+    private static AuthenticationSessionModel createOrJoinLogoutSession(KeycloakSession session, RealmModel realm, final AuthenticationSessionManager asm, UserSessionModel userSession, boolean browserCookie) {
         // Account management client is used as a placeholder
         ClientModel client = realm.getClientByClientId(Constants.ACCOUNT_MANAGEMENT_CLIENT_ID);
 
-        AuthenticationSessionModel logoutAuthSession = asm.getCurrentAuthenticationSession(realm, client);
-        // Try to join existing logout session if it exists and browser session is required
-        if (browserCookie && logoutAuthSession != null) {
-            if (Objects.equals(AuthenticationSessionModel.Action.LOGGING_OUT.name(), logoutAuthSession.getAction())) {
-                return logoutAuthSession;
-            }
-            // Re-create the authentication session for logout
-            logoutAuthSession = logoutAuthSession.getParentSession().createAuthenticationSession(client);
+        // Try to lookup current authSessionId from browser cookie. If doesn't exists, use the same as current userSession
+        String authSessionId = null;
+        boolean browserCookiePresent = false;
+        if (browserCookie) {
+            authSessionId = asm.getCurrentAuthenticationSessionId(realm);
+        }
+        if (authSessionId != null) {
+            browserCookiePresent = true;
         } else {
-            RootAuthenticationSessionModel rootLogoutSession = asm.createAuthenticationSession(realm, browserCookie);
-            logoutAuthSession = rootLogoutSession.createAuthenticationSession(client);
+            authSessionId = userSession.getId();
+        }
+
+        // Try to join existing logout session if it exists
+        RootAuthenticationSessionModel rootLogoutSession = session.authenticationSessions().getRootAuthenticationSession(realm, authSessionId);
+        if (rootLogoutSession == null) {
+            rootLogoutSession = session.authenticationSessions().createRootAuthenticationSession(authSessionId, realm);
         }
+        if (browserCookie && !browserCookiePresent) {
+            // Update cookie if needed
+            asm.setAuthSessionCookie(authSessionId, realm);
+        }
+
+        // See if we have logoutAuthSession inside current rootSession. Create new if not
+        Optional<AuthenticationSessionModel> found = rootLogoutSession.getAuthenticationSessions().values().stream().filter((AuthenticationSessionModel authSession) -> {
+
+            return client.equals(authSession.getClient()) && Objects.equals(AuthenticationSessionModel.Action.LOGGING_OUT.name(), authSession.getAction());
+
+        }).findFirst();
+
+        AuthenticationSessionModel logoutAuthSession = found.isPresent() ? found.get() : rootLogoutSession.createAuthenticationSession(client);
```

Comment:
nit: can be replaced with
```java
   AuthenticationSessionModel logoutAuthSession = found.orElse(rootLogoutSession.createAuthenticationSession(client));
```

Reply:
Thanks Hynek for the review! For this one, I remember I wanted to do something like this when I was coding it, but it requires "rootLogoutSession" to be declared as "final" variable, which it isn't, so would need to be re-declared. So did it the current way in the end. Will doublecheck again and possibly change it...

Reply:
@hmlnarik Sorry, just tried that it works when I do it exactly the way you mentioned. I've used a bit different syntax before, which requires "rootLogoutSession" to be final :) Changed it in the other PR #4837 .